### PR TITLE
Fix support for complex parameters of parametric aggregate functions, #30975

### LINKED ISSUE #30975: Exception thrown when using sumMapFiltered on a distributed table
**Describe what's wrong**

An exception is thrown when trying to use the `sumMapFiltered` function on a distributed table. 

**Does it reproduce on recent release?**

Yes. The tests below ran on ClickHouse 21.9.3

**How to reproduce**

Create the test tables and insert a couple rows of data:
```sql
CREATE TABLE local ON CLUSTER 'cluster_name' (
  time     DateTime,
  data     Nested(
    key      Int16,
    value    Int64
  )
) ENGINE = MergeTree()
  PARTITION BY toYYYYMM(time)
  ORDER BY time

CREATE TABLE distributed ON CLUSTER 'cluster_name' (
  time     DateTime,
  data     Nested(
    key      Int16,
    value    Int64
  )
) ENGINE = Distributed('cluster_name', 'cluster_name', 'local') 

INSERT INTO local VALUES
  (1635794296, [1, 2, 3], [10, 10, 10]),
  (1635794296, [1, 2, 3], [10, 10, 10]),
  (1635794296, [1, 2, 3], [10, 10, 10])
```

Then run this query against the distributed table:

```sql
SELECT sumMapFiltered([toInt16(1),toInt16(2)])(data.key, data.value) FROM distributed
```

**Expected behavior**

I would expect the query to return the correct aggregation value, but an error is thrown instead

**Error message and/or stacktrace**

```
Received exception from server (version 21.9.3):
Code: 62. DB::Exception: Received from localhost:9000. DB::Exception: Syntax error (data type): failed at position 34 ('['): [1, 2]), Array(Int16), Array(Int64)). Expected one of: number, nested table, literal, NULL, identifier, data type argument, string literal, name and type pair list, list of elements, data type, list, delimited by binary operators, name and type pair: while receiving packet from [redacted]:9000: While executing Remote. (SYNTAX_ERROR)
```

**Additional context**

The exact same query works as expected when executed against the local table.

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix support for complex parameters (like arrays) of parametric aggregate functions. This closes #30975. The aggregate function `sumMapFiltered` was unusable in distributed queries before this change.

### CODE REVIEW

File: src/AggregateFunctions/AggregateFunctionSumMap.h [hunk 1]
```diff
@@ -495,14 +496,15 @@ class AggregateFunctionSumMapFiltered final :
                 "Aggregate function {} requires an Array as a parameter",
                 getName());
 
+        this->parameters = params_;
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/AggregateFunctionSumMap.h:499]: This is the first fix - parameters were lost in the data type.

File: src/Parsers/ParserDataType.cpp [hunk 2]
```diff
@@ -27,7 +27,7 @@ class ParserDataTypeArgument : public IParserBase
     {
         ParserNestedTable nested_parser;
         ParserDataType data_type_parser;
-        ParserLiteral literal_parser;
+        ParserAllCollectionsOfLiterals literal_parser;
```
Comment (alexey-milovidov) [inline src/Parsers/ParserDataType.cpp:30]: And this is the second fix to allow complex parameters.

Comment (evillique) [reply src/Parsers/ParserDataType.cpp:30]: One interesting thing about `ParserAllCollectionsOfLiterals` is that we would also parse expressions like `{1:'a'}` approximately as `Map(UInt8, String)`, which is not what we generally do.

Also, this parser is a bit simpler than others in the sense that we do not fill the `begin` and `end` fields of ASTLiteral, which is used at least in one place in the code. (src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp:199)

I will make the fix for it soon.

**Review by evillique — APPROVED** [review]

File: src/Parsers/ParserDataType.cpp (hunk 2 above)
Comment (evillique) [inline src/Parsers/ParserDataType.cpp:30]: After an update to the latest master, this change should be applied.

It is also worth noticing that ParserAllCollectionsOfLiterals does not parse literals (as in ParserLiteral), but only collections of literals.
```suggestion
        ParserAllCollectionsOfLiterals literal_parser(false);
```

**Pull request merged**